File: src/net/eduard/api/server/kit/KitAbility.java
package net.eduard.api.server.kit;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.inventory.ItemStack;

import net.eduard.api.lib.modules.Mine;
import net.eduard.api.lib.click.PlayerInteract;
import net.eduard.api.lib.manager.CooldownManager;
import net.eduard.api.lib.modules.KitType;
import net.eduard.api.lib.storage.StorageAttributes;

public class KitAbility extends CooldownManager {

	private String name;
	private double price;
	private boolean showOnGui = true;
	private boolean enabled = true;
	private String disabled = "§6Habilidade desativada temporariamente!";
	private boolean activeCooldownOnPvP;
	private int times = 1;
	private ItemStack icon;
	@NotCopyable
	private transient PlayerInteract click;
	@NotCopyable
	private transient Map<Player, Integer> timesUsed = new HashMap<>();
	@StorageAttributes(reference = true)
	private List<KitAbility> kits = new ArrayList<>();
	private transient List<Player> players = new ArrayList<>();

	

	public List<KitAbility> getKits() {
		return kits;
	}

	public void setKits(List<KitAbility> kits) {
		this.kits = kits;
	}

	public KitAbility() {
		this("", KitType.DEFAULT);

	}

	public KitAbility(String name) {
		this(name, KitType.DEFAULT);

	}

	public KitAbility(String name, KitType type) {
		if (name.isEmpty()) {
			setName(getClass().getSimpleName());
		} else {
			setName(name);
		}
		REQUIRE_PERMISSION = name.toLowerCase().replace(" ", "");
	}

	public ItemStack add(ItemStack item) {
		GIVE_ITEMS.add(Mine.setName(item, "§b" + name));
		return item;
	}

	public ItemStack add(Material material) {
		return add(material, 0);
	}

	public ItemStack add(Material material, int data) {
		return add(new ItemStack(material, 1, (short) data));
	}

	@Override
	public boolean cooldown(Player player) {
		if (!enabled) {
			Mine.send(player, disabled);
			return false;
		}
		if (onCooldown(player)) {
			sendOnCooldown(player);
			return false;
		}
		int x = 0;
		if (timesUsed.containsKey(player)) {
			x = timesUsed.get(player);
		}
		x++;
		if (x == times) {
			setOnCooldown(player);
			sendStartCooldown(player);
			timesUsed.remove(player);
		} else {
			timesUsed.put(player, x);
		}
		return true;
	}

	@EventHandler
	public void event(EntityDamageByEntityEvent e) {
		if (activeCooldownOnPvP) {
			if (e.getEntity() instanceof Player) {
				Player p = (Player) e.getEntity();
				if (hasKit(p)) {
					setOnCooldown(p);
				}
			}
		}
	}

	public ItemStack getIcon() {
		return icon;
	}

	public String getName() {
		return name;
	}

	public List<Player> getPlayers() {
		return players;
	}

	public double getPrice() {
		return price;
	}

	public int getTimes() {
		return times;
	}

	public boolean hasKit(Player player) {
		return players.contains(player);
	}

	public boolean isActiveCooldownOnPvP() {
		return activeCooldownOnPvP;
	}

	public KitAbility setActiveCooldownOnPvP(boolean activeCooldownOnPvP) {
		this.activeCooldownOnPvP = activeCooldownOnPvP;
		return this;
	}

	public KitAbility setIcon(ItemStack icon) {
		this.icon = icon;
		return this;
	}

	public KitAbility setIcon(Material material, int data, String... lore) {
		icon = new ItemStack(material);
		Mine.setName(icon, "§6Kit " + name);
		Mine.setLore(icon, lore);
		Mine.addEnchant(icon, Enchantment.DURABILITY, 10);
		return this;
	}

	public KitAbility setIcon(Material material, String... lore) {
		return setIcon(material, 0, lore);
	}

	public KitAbility setName(String name) {
		this.name = name;
		return this;
	}

	public KitAbility setPlayers(ArrayList<Player> players) {
		this.players = players;
		return this;
	}

	public KitAbility setPrice(double price) {
		this.price = price;
		return this;
	}

	public KitAbility setTimes(int times) {
		this.times = times;
		return this;
	}

	public boolean isShowOnGui() {
		return showOnGui;
	}

	public void setShowOnGui(boolean showOnGui) {
		this.showOnGui = showOnGui;
	}

	public boolean isEnabled() {
		return enabled;
	}

	public void setEnabled(boolean enabled) {
		this.enabled = enabled;
	}

	public String getDisabled() {
		return disabled;
	}

	public void setDisabled(String disabled) {
		this.disabled = disabled;
	}

	public PlayerInteract getClick() {
		return click;
	}

	public void setClick(PlayerInteract click) {
		this.click = click;
	}

}

File: src/net/eduard/api/server/kit/KitAbility.kt
package net.eduard.api.server.kit

import java.util.ArrayList
import java.util.HashMap

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.inventory.ItemStack

import net.eduard.api.lib.modules.Mine
import net.eduard.api.lib.click.PlayerInteract
import net.eduard.api.lib.game.Explosion
import net.eduard.api.lib.game.Jump
import net.eduard.api.lib.manager.CooldownManager
import net.eduard.api.lib.manager.EffectManager
import net.eduard.api.lib.manager.EventsManager
import net.eduard.api.lib.modules.BukkitTimeHandler
import net.eduard.api.lib.storage.Storable
import org.bukkit.plugin.Plugin

open class KitAbility : EventsManager(), BukkitTimeHandler {

    @Transient
    var click: PlayerInteract? = null

    @Transient
    val timesUsed = HashMap<Player, Int>()
    var name = "Kit"
    var price = 0.0
    var isShowOnGui = true
    var isEnabled = true
    var activeCooldownOnPvP = false
    var useLimit = 1
    val lore = ArrayList<String>()
    var itemType = Material.DIAMOND_AXE
    var itemData = 0
    var effects = EffectManager()
    var jump: Jump? = null
    var explosion: Explosion? = null
    var cooldown = CooldownManager()
    val permission get() = effects.requirePermission


    val potions
        get() = effects.potionsToApply

    var time: Int
        set(value) {
            cooldown.duration = value * 20.toLong()
        }
        get() = cooldown.duration.toInt()


    var sound
        get() = effects.soundToPlay
        set(value) {
            effects.soundToPlay = value
        }


    var display
        get() = effects.visualEffectToShow
        set(value) {
            effects.visualEffectToShow = value
        }


    val items
        get() = effects.itemsToGive


    var message
        get() = effects.messageToSend
        set(value) {
            effects.messageToSend = value
        }

    val icon: ItemStack
        get() {

            val item = ItemStack(itemType, 1)
            val meta = item.itemMeta
            meta.displayName = "§e$name"
            val list = mutableListOf<String>()
            for (line in lore) {
                list.add("§f$line")
            }
            meta.lore = list
            item.itemMeta = meta


            return item
        }

    @Storable.StorageAttributes(reference = true)
    var kits: List<KitAbility> = ArrayList()

    @Transient
    var players: List<Player> = ArrayList()

    init {

        if (name.isEmpty() || name == "Kit")
            name = (javaClass.simpleName)


        effects.requirePermission = name.toLowerCase().replace(" ", "")
    }

    fun add(item: ItemStack) {
        items.add(Mine.setName(item, "§b" + name))

    }

    fun add(type: Material) {
        items.add(Mine.newItem(type, "§b$name"))

    }


    fun onCooldown(player: Player): Boolean {
        return cooldown.onCooldown(player)
    }

    fun setOnCooldown(player: Player) {
        cooldown.setOnCooldown(player)
    }

    fun cooldown(player: Player): Boolean {
        if (!isEnabled) {
            Mine.send(player, "")
            return false
        }
        if (cooldown.onCooldown(player)) {
            cooldown.sendOnCooldown(player)
            return false
        }
        var x = 0
        if (timesUsed.containsKey(player)) {
            x = timesUsed.getOrDefault(player, 0)
        }
        x++
        if (x == useLimit) {
            cooldown.setOnCooldown(player)
            cooldown.sendStartCooldown(player)
            timesUsed.remove(player)
        } else {
            timesUsed[player] = x
        }
        return true
    }

    @EventHandler
    open fun event(e: EntityDamageByEntityEvent) {
        if (activeCooldownOnPvP) {
            if (e.entity is Player) {
                val p = e.entity as Player
                if (hasKit(p)) {
                    cooldown.setOnCooldown(p)
                }
            }
        }
    }


    fun hasKit(player: Player): Boolean {
        return players.contains(player)
    }


    fun setIcon(material: Material, vararg lore: String) {
        setIcon(material, 0, *lore)
    }

    fun setIcon(material: Material, data: Int, vararg lore: String) {
        itemType = material
        itemData = data
        this.lore.addAll(lore)
    }

    override fun getPluginConnected(): Plugin {
        return plugin
    }


}
